edb/edgeql/compiler/inference/cardinality.py

```diff
@@ -932,6 +936,82 @@ def extract_filters(
     return None
 
 
+def _get_object_constraint_pointers(
```

Comment:
Hm, I did a bunch of work to only look at tuple constraints so that everything is pretty regular and works without any trouble. But now that I think about it, maybe that's all a waste of time and an unnecessary limitation: any constraint that is a stable function of the pointers it references should work. So maybe we should drop all of this logic and `irutils.extract_constraint_style_ptrs` and *just* base the decision on `subjectexpr.refs` (which we already look at as an optimization before looking more deeply at the subjectexpr body).

The advantage here would be that then this would work with constraints on stuff like `.first ++ .last` would work. To the extent that that is really an advantage, I guess.

Thoughts?

Reply:
This would work only if the constraint expression result is not _more_ unique than a tuple of its inputs, which would be hard to prove in the general case.  To support inference on arbitrary expressions, we must verify that you've used an equivalent expression in your `FILTER` or `UNLESS CONFLICT`.

Reply:
Hm, why is being "more unique" a problem? The max cardinality is still ONE?

Reply:
```
function foo(first_name: str, last_name: str) -> str {
    USING (
        SELECT 
            first_name 
            ++ last_name 
            ++ <str>count((
                SELECT User 
                FILTER (.first_name, .last_name) = (first_name, last_name)
            ))
    )
}

type Foo { 
    property first_name -> str; 
    property last_name -> str; 
    constraint exclusive on (foo(.first_name, .last_name));
}
```

That said, as I was writing the above, I remembered that we shouldn't allow non-`Immutable` functions in constraint expressions, so this issue is probably moot.

Reply:
Oh, yeah; I said stable above but should have said immutable.

With that, do you think it's worth making this simplification/change?

Reply:
Let's do it in a separate PR.

Reply:
Sure. I think that PR will end up being "delete half of this PR"